integration_tests/auction_test.go

```diff
@@ -0,0 +1,256 @@
+			SaleTokenUnitPriceInUsomm: sdk.MustNewDecFromStr("2"),
+			TotalUsommPaid:            sdk.NewCoin("usomm", sdk.NewIntFromUint64(5000)),
+		}
+
+		s.T().Log("Verifying bid stored as expected.")
+		actualBid, err := auctionQueryClient.QueryBid(context.Background(), &types.QueryBidRequest{BidId: uint64(1), AuctionId: uint32(1)})
+		s.Require().NoError(err)
+		s.Require().Equal(expectedBid, *actualBid.Bid)
+		s.T().Log("Bid stored correctly!")
+
+		s.T().Log("Submitting another bid...")
+		// Submit another bid to be partially fulfilled
+		bidRequest2 := types.MsgSubmitBidRequest{
+			AuctionId:              uint32(1),
+			Bidder:                 orch0Address,
+			MaxBidInUsomm:          sdk.NewCoin("usomm", sdk.NewIntFromUint64(10000)),
+			SaleTokenMinimumAmount: sdk.NewCoin("gravity0x3506424f91fd33084466f402d5d97f05f8e3b4af", sdk.NewIntFromUint64(50)),
+			Signer:                 orch0Address,
+		}
+
+		_, err = s.chain.sendMsgs(*orchClientCtx, &bidRequest2)
+		s.Require().NoError(err)
+		s.T().Log("Bid submitted successfully!")
+
+		s.T().Log("Verifying user funds debited and credited appropriately.")
+		// Verify user has funds debited and purchase credited
+		balanceRes, err = bankQueryClient.AllBalances(context.Background(), &banktypes.QueryAllBalancesRequest{Address: orch0Address})
+		s.Require().NoError(err)
+		s.T().Logf("Orchestrator 0 token balances after second bid %v", balanceRes.Balances)
+
+		s.Require().Equal(sdk.NewCoin("gravity0x3506424f91fd33084466f402d5d97f05f8e3b4af", sdk.NewInt(5000)), balanceRes.Balances[0])
+		s.Require().Equal(sdk.NewCoin("usomm", initialOrchBalanceRes.Balances[1].Amount.Sub(sdk.NewInt(10000))), balanceRes.Balances[2])
+		s.T().Log("User funds updated correctly!")
+
+		s.T().Log("Verifying auction has completed...")
+		// Verify no active auctions
+		_, err = auctionQueryClient.QueryActiveAuctions(context.Background(), &types.QueryActiveAuctionsRequest{})
+		s.Require().Error(err)
+		s.Require().Equal("rpc error: code = NotFound desc = rpc error: code = NotFound desc = No active auctions found: key not found", err.Error())
+		s.T().Log("Auction completed successfully!")
+
+		s.T().Log("Verifying ended auction stored correctly..")
+		// Verify ended auction
+		endedAuctionResponse, err := auctionQueryClient.QueryEndedAuction(context.Background(), &types.QueryEndedAuctionRequest{AuctionId: uint32(1)})
+		s.Require().NoError(err)
+
+		node, err := orchClientCtx.GetNode()
+		s.Require().NoError(err)
+		status, err := node.Status(context.Background())
+		s.Require().NoError(err)
+		currentBlockHeight := status.SyncInfo.LatestBlockHeight
+
+		expectedEndedAuction := types.Auction{
+			Id:                         uint32(1),
+			StartingTokensForSale:      sdk.NewCoin("gravity0x3506424f91fd33084466f402d5d97f05f8e3b4af", sdk.NewInt(5000)),
+			StartBlock:                 uint64(0),
+			EndBlock:                   uint64(currentBlockHeight),
```

Comment:
I don't think I've seen this test fail in the 10-20 times I've run it because the code seems to run fast enough, but if we want to err on the side of caution we can do a piecewise field check here and assert the expected end block is >= the actual one. Thoughts?

Reply:
That might be a bit much considering it's comparing the whole object, if we observe flakiness we may want to change.